src/Generating/FinishGen.cpp

```diff
@@ -665,51 +665,106 @@ void cFinishGenVines::GenFinish(cChunkDesc & a_ChunkDesc)
 
 
 
-
-
 ////////////////////////////////////////////////////////////////////////////////
 // cFinishGenSprinkleFoliage:
 
-bool cFinishGenSprinkleFoliage::TryAddSugarcane(cChunkDesc & a_ChunkDesc, int a_RelX, int a_RelY, int a_RelZ)
+bool cFinishGenSprinkleFoliage::TryAddCactus(cChunkDesc &a_ChunkDesc, int a_RelX, HEIGHTTYPE & a_RelY, int a_RelZ)
 {
-	// We'll be doing comparison to neighbors, so require the coords to be 1 block away from the chunk edges:
-	if (
-		(a_RelX < 1) || (a_RelX >= cChunkDef::Width  - 1) ||
-		(a_RelY < 1) || (a_RelY >= cChunkDef::Height - 2) ||
-		(a_RelZ < 1) || (a_RelZ >= cChunkDef::Width  - 1)
-	)
+	if (!IsDesertVariant(a_ChunkDesc.GetBiome(a_RelX, a_RelZ)))
 	{
 		return false;
 	}
 
-	// Only allow dirt, grass or sand below sugarcane:
-	switch (a_ChunkDesc.GetBlockType(a_RelX, a_RelY, a_RelZ))
+	for (int i = 0; i < 1 + (m_Noise.IntNoise2DInt(a_RelX, a_RelZ) % m_MaxCactusHeight); i++)
 	{
-		case E_BLOCK_DIRT:
-		case E_BLOCK_GRASS:
-		case E_BLOCK_SAND:
+		if (
+			(a_RelX < 1) || (a_RelX >= cChunkDef::Width - 1) ||
+			(a_RelY < 1) || (a_RelY >= cChunkDef::Height - 2) ||
+			(a_RelZ < 1) || (a_RelZ >= cChunkDef::Width - 1)
+		)
 		{
-			break;
+			return (i == 0) ? false : true;
 		}
-		default:
+
+		switch (a_ChunkDesc.GetBlockType(a_RelX, a_RelY, a_RelZ))
 		{
-			return false;
+			case E_BLOCK_SAND :
+			case E_BLOCK_CACTUS :
+			{
+				break;
+			}
+			default :
+			{
+				return (i == 0) ? false : true;
+			}
 		}
+
+		int y = a_RelY + 1;
+		if ((
+			!(a_ChunkDesc.GetBlockType(a_RelX + 1, y, a_RelZ)     == E_BLOCK_AIR) ||
+			!(a_ChunkDesc.GetBlockType(a_RelX - 1, y, a_RelZ)     == E_BLOCK_AIR) ||
+			!(a_ChunkDesc.GetBlockType(a_RelX,     y, a_RelZ + 1) == E_BLOCK_AIR) ||
+			!(a_ChunkDesc.GetBlockType(a_RelX,     y, a_RelZ - 1) == E_BLOCK_AIR))
```

Comment:
A cactus can also be placed near some non solid blocks:
![2018-02-09_12 02 33](https://user-images.githubusercontent.com/1812304/36024844-4ca20ec0-0d91-11e8-918c-79eea8444d50.png)

Based on my wiki these are:

- Sugarcane
- Lever
- Button
- Flower
- Sapling
- Armor Stand
- Torch
- Carpets
- Rails
- Snow
- Repeater
- Comparator

Maybe only sugarcane could create a problem, but I would suggest creating a function to check for the nearby block, something like `BlockInfo::CanBePlacedAgainstCactus`. Maybe the code handling the cactus destructon on block place already has such a function ❓